### PR TITLE
Feat/populate-dev

### PR DESCRIPTION
**Pull request by jiixing**
A script (scripts/populate-runner.js) to populate empty database via .json files in scripts/dev-data.  The code is smart enough to not reinsert duplicate row, so it can also be used to add additional data to the database.

Currently only 4 tables  are populated, I will add more tables in the future as I test them

populate-runner.js
```
const tables = ['user','appointment','message','question']
```

### CODE REVIEW

File: src/scripts/populate-dev.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
+const FS = require("q-io/fs");
+import mongooseConf from '../config/mongoose.conf.js';
+import {validateEnvVariables} from '../config/env.conf.js';
+import mongoose from 'mongoose';
+import _ from 'lodash';
+
+function getKlass(tableName) {
```
Comment (binarious) [inline src/scripts/populate-dev.js]: What's `Klass`?

File: src/scripts/populate-dev.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    let klass = require(`../app/models/${tableName}.model`)
```
Comment (binarious) [inline src/scripts/populate-dev.js]: Please stick to the indentation with two spaces.

File: src/scripts/populate-dev.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    klass = klass.default // import default equivalent
+    if (!klass)
+        throw new Error(`file definition ../app/models/${tableName}.model not found`)
+    return klass
+}
+
+async function saveRow(klass, data) {
+    // find duplicate
+    let clean_data = _.omit(data, _.keys(mapping))
+    // console.log("clean_data",clean_data)
```
Comment (binarious) [inline src/scripts/populate-dev.js]: Please remove this.

File: src/scripts/populate-dev.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    let existing = await klass.findOne(clean_data)
+
+    if (!existing) {
+
```
Comment (binarious) [inline src/scripts/populate-dev.js]: Please remove those empty lines at the beginning of blocks.

File: src/scripts/populate-dev.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        let find_err = false
+
+        for (let k of keys) {
+            let res = await mapping[k](data[k])
+
+            if (res) {
+                //collect find_result
+                find_result = Object.assign(find_result, res)
+            }
+            else {
```
Comment (binarious) [inline src/scripts/populate-dev.js]: Please put `else` on the same line as `}`.

File: src/scripts/populate-dev.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+function init() {
+    validateEnvVariables();
+    mongooseConf(mongoose);
+}
+
+
+/**
+ * convert map-user to user.id
+ */
+let user_id_functor = async (data) => {
```
Comment (binarious) [inline src/scripts/populate-dev.js]: Please stick to camelCased variable names.

File: src/scripts/populate-dev.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+ * @returns {Promise.<void>}
+ */
+module.exports = async function start(tables) {
+
+    init()
+    //  await test() // when developing new mapping functor
+    for (let t of tables) {
+        await readJson(t, `./dev-data/${t}.json`)
+    }
+    /*
```
Comment (binarious) [inline src/scripts/populate-dev.js]: Remove unused code

**Review by binarious — CHANGES_REQUESTED** [review]
Good idea :+1:. Some style problems.

File: src/scripts/populate-dev.js (hunk 7 above)
Comment (jiixing) [reply src/scripts/populate-dev.js]: I fixed the file as requested.

klass is renamed to model. It is a way to programmaticly import model.js module, eg 'user.model.js'

**Review by binarious — DISMISSED** [review]

**Comment by binarious** [comment]
LGTM and would make things easier. What do you think @olastor?

**Comment by jiixing** [comment]
I 'git pull --rebase'  on the updated master branch so the travis checks passed

File: package.json [hunk 8]
```diff
@@ -16,6 +16,7 @@
   "dependencies": {
     "bcrypt-nodejs": "0.0.3",
     "body-parser": "^1.17.1",
+    "colored-console":"^0.10.0",
```
Comment (olastor) [inline package.json]: Can you please use a package that is more popular and better maintained? Like https://www.npmjs.com/package/colors seems to be a better choice here.

**Review by olastor — CHANGES_REQUESTED** [review]

File: package.json (hunk 8 above)
Comment (jiixing) [reply package.json]: I tried colors again, and it didn't work in Webstorm terminal see https://github.com/Marak/colors.js/issues/196, another issue is it doesn't display Object member for better output.  Attached is comparison display with colors on the right. I have the changes in stashed if you still want to use colors

colored-console

Comment (olastor) [reply package.json]: I think you could use the logger we already have at https://github.com/Sirimangalo/meditation-plus-server/blob/master/src/app/helper/logger.js . It should also have different colors (more about usage here https://github.com/winstonjs/winston#using-logging-levels).

**Comment by jiixing** [comment]
I am using 'winston' console logger as colorized log now

File: src/scripts/populate-dev.js [hunk 9]
```diff
@@ -1,10 +1,20 @@
 import FS from'q-io/fs';
-import cconsole  from 'colored-console';
+import Winston  from 'winston';
```
Comment (olastor) [inline src/scripts/populate-dev.js]: Is there a reason for declaring a new logger instead of `import { logger } from '../helper/logger.js;'`?

**Comment by jiixing** [comment]
@olastor Sorry didn't read your previous comment fully, I was only focusing on using winston. It is using app/helper/logger now

**Review by olastor — DISMISSED** [review]

**Comment by olastor** [comment]
@jiixing No problem. Thanks for the quick changes!

File: package.json [hunk 10]
```diff
@@ -16,7 +16,6 @@
   "dependencies": {
     "bcrypt-nodejs": "0.0.3",
     "body-parser": "^1.17.1",
-    "colored-console":"^0.10.0",
```
Comment (binarious) [inline package.json]: Please rerun `yarn` so that `colored-console` will be removed from the `yarn.lock` file.

**Review by binarious — CHANGES_REQUESTED** [review]

**Comment by jiixing** [comment]
@binarious  I updated package.json and yarn.lock from master, and then 'yarn add q-io --dev' that I use

**Review by binarious — APPROVED** [review]

**Comment by binarious** [comment]
Thank you, @jiixing.

**Pull request merged**